### ISSUE TITLE
How to redirect to data.additionalData.url on click in foreground / background

### ISSUE THREAD

**Issue by fusionsquare** [phonegap/phonegap-plugin-push#81]
Hi,

great update! It works very well. But how can i open a URL that i send with the notification?

push.on('notification', function(data) {
            if( data.additionalData.url ) {
                window.location.href = data.additionalData.url;
            }
        });

That works when the app are closed. The app opens and redirect to the specific url that i send within the notification. 

But how can i redirect the user when he clicks on the notification message if the App are in background?

I had the same issue with the old plugin too.

**Comment by mv83** [comment]
I have the same problem. I'm following...

**Comment by macdonst** [comment]
Doesn't this work for you:

```
        push.on('notification', function(data) {
            if (data.additionalData.url) {
                console.log("URL: " + data.additionalData.url);
                document.location.href = data.additionalData.url;
            }
        });
```

**Comment by mv83** [comment]
Yes It works if the app is in foreground. But what about if the app is in background?

**Comment by macdonst** [comment]
@mv83 if the app is in the background or stopped then nothing will happen but if the users clicks on the notification in the shade it should launch the app, the notification should be fired, then your app gets redirected.

Just tested on Android with latest plugin. What behaviour are you seeing?

**Comment by fusionsquare** [comment]
Here a overview what works and whats not:

> app closed and send a notification:
working - push comes in,  on click at the push notification windowon click the app opens and the additionalData.url will be loaded

> app on foreground (remote website loaded already) and send a notification:
not working - nothings happend, no push incoming, not sound, nothing

> app on background (the remote website is opened):
not working - push comes in, on click at the push notification window the app opened and shows the last url from the remote website

I mean, it can be that the Push Plugin dont have the enviroment from cordova / phonegap at the remote website? On "app is closed" everything works, after that the redirect to the remote website it doesnt.

**Comment by macdonst** [comment]
@fusionsquare wait are you saying your app loads a remote page that does not include cordova.js? If that is the case all of your JS from the app will be unloaded and your on "notification" handler will never be executed.

If your remote website is not a PhoneGap/Cordova app then none of the PhoneGap/Cordova API's will work.

**Comment by fusionsquare** [comment]
@macdonst exactly. I try'd but it doesn't work. Can you tell me how to implement the files and wich files are needed? A link oder hint will be awesome. I have this problems since february and don't find any solution for this. Hope you can help!

**Comment by mv83** [comment]
@macdonst thanks for your reply. Yes, when I click on the notification, the app is launched but the last page visited is showed. I would always open the same page after clicking the notification. Is it possible?

**Comment by fusionsquare** [comment]
@mv83 i think the problem is, that the app doesn't trigger the resume. Did you work with a remote website too?

**Comment by mv83** [comment]
@macdonst no, I'm developing the entire project with phonegap/cordova and all pages are in the sub-folder www of the folder assets...

**Comment by macdonst** [comment]
When you click on the push notification the app will be launched. If the app is not running it will be started and your main page will be shown. If the app is in the background then it will be resumed but it will display whatever the last page you were viewing.

If a url is not provided you could just redirect back to the main page:

```
push.on('notification', function(data) {
  if (data.additionalData.url) {
    console.log("URL: " + data.additionalData.url);
    document.location.href = data.additionalData.url;
  } else {
    document.location.href = "index.html";
  }
});
```

@mv83 I believe your problem is when you redirect to a sub page you have unloaded the push notification handler. I'm not sure who your app is architected but you might want to look at single page app patterns.

 @fusionsquare You'd have to host the cordova.js and plugin files at your remote site. Why do you have things setup this way?

**Comment by fusionsquare** [comment]
@macdonst Ok, i try. Can't find the plugin files, where can i get them?

**Comment by jarnovanrhijn** [comment]
@macdonst, it has been a year since the last comment on this issue and i was wondering if it is possible to catch the event when resuming from a notification?

**Comment by GaneshSawant** [comment]
Hello @fusionsquare , 

I am trying to achieve similar functionality  and currently using hybrid Remote approach for building android app.
May I know if the above logic applicable for hybrid remote approach. 
I tried to add the same logic inside com.salesforce.util.push.js but it didnt work. 
Am I missing something here ?

Appreciate your prompt response.

Thank you,

**Comment by yesroh** [comment]
Boys ( @fusionsquare and @GaneshSawant ) I am trying to achieve similar functionality and currently using hybrid Remote approach for building android app.

I have put my logic inside the event:
   
      $rootScope.$on('hide:notification', (event, data) => {
        console.log('olá, estou aqui!');
        $mdDialog.hide();
      });

but the $mdDialog.hide(); don't is fired. there is a way of doing this?

**Comment by mountainclimbing** [comment]
Hi guys!
I'm using cordova and the notification part for FCM is like this

push.on('notification', function(data) {
		  if (data.additionalData.url) {
                console.log("url: " + data.additionalData.url);
                document.location.href = data.additionalData.url;
            }else {
	$.mobile.changePage("#homepage");
  }

});

The URL part, is it posted separately from the server side or it has to be in the message body?

By the way, on the server side, how have you handled "'click_action' => 'FCM_PLUGIN_ACTIVITY'" is it commented out or its active?
See
    $msg = array (
		'body'  => 'we ned it',
        'title' => $titles,
		'url' => 'http://google.com/',
        'click_action' => 'FCM_PLUGIN_ACTIVITY',
    );
